src/core/thread/mlr_manager.cpp

```diff
@@ -117,6 +117,50 @@ void MlrManager::UpdateLocalSubscriptions(void)
     LogMulticastRegistrationTable();
 }
 
+#if OPENTHREAD_FTD
+void MlrManager::UpdateProxiedSubscriptions(const Child &aChild)
+{
+    uint16_t                  childIndex = Get<ChildTable>().GetChildIndex(aChild);
+    Child::Ip6AddressIterator childAddrIter;
+    Ip6::Address              address;
+
+    // Clear deprecated multicast address of this Child
+    for (MulticastListenerRegistrationTable::Iterator iter(GetInstance(),
+                                                           MulticastListenerRegistrationTable::kFilterProxied);
+         !iter.IsDone(); iter.Advance())
+    {
+        MulticastListenerRegistration &registration = iter.GetMulticastListenerRegistration();
+        bool                           proxied;
+
+        if (!registration.IsChildProxied(childIndex))
+        {
+            continue;
+        }
+
+        proxied = aChild.IsStateValidOrRestoring() && aChild.HasIp6Address(registration.GetAddress());
+        ot::MulticastListenerRegistrationTable::SetChildProxied(registration, childIndex, proxied);
+    }
+
+    VerifyOrExit(aChild.IsStateValidOrRestoring(), OT_NOOP);
+
+    // Add new multicast address of this Child
+    while (aChild.GetNextIp6Address(childAddrIter, address) == OT_ERROR_NONE)
+    {
+        if (!address.IsMulticastLargerThanRealmLocal())
+        {
+            continue;
+        }
+
+        mTable.SetChildProxied(address, childIndex, true);
```

Comment:
Continuing the conversation from https://github.com/openthread/openthread/pull/5219#discussion_r455473665:

> As for the proxied MLR.req, we analyzed the logic of sending MLR.req without using any table and found that the time complexity could be O(ChildNum * IpNumPerChild * IP6AddressesTlvSize). On the other hand, the current implementation using a table has a time complexity of O(MulticastIpNum). I think we can discuss more in the proxied MLR.req PR #5236 .

I think there are ways to reduce the overhead. For example, we could maintain a bit-vector with each bit corresponding to a child's address in the child table. A bit is set if the corresponding child's address is to be used when sending an MLR.req.

When adding a child multicast address, check to see if any of the addresses marked for use in MLR.req match (i.e. the address entries with the corresponding bit set).

1. If there is a match, clear the bit corresponding to the address being added. This indicates that the newly added address is already included in MLR.req messages.

1. If there is no match, set the bit corresponding to the address being added. This indicates that the newly added address needs to be added to MLR.req messages.

When removing a child multicast address:

1. If the corresponding bit is set, check to see if any other child has a matching address. If no matching address, clear the bit. This indicates that the address can be removed from the MLR.req message.

1. If the corresponding bit is clear, simply remove the address as this indicates that other children are still using the address.

When sending an MLR.req message, simply send the multicast addresses that have the bit set.

If we wanted to further optimize memory, then we could distribute the bit vector into the child table - I believe there is some padding that we could use to avoid any additional memory overhead. If we wanted to avoid scanning each child table entry, we could have a top-level bit vector that indicates whether a child entry has any such bit set.

Thoughts?

Reply:
Yes we only need to check Child addresses that was marked as newly added.
We have a similar flag defined in #5236 : `MlrState` whose value can be `ToRegister (0)` or `Registering (1)` or `Registered (2)`

I have created PR #5248 which has less RAM cost and code size (to my supprise). 
The reduction in code size was mainly due to the compaction of `MlrManager::UpdateLocalSubscriptions` because it does not need to copy addresses anymore.

Reply:
@jwhui @librasungirl @abtink 
Some further thoughts on this:

Let's first assume that Child has a flag for each Address (domain >= 4) to indicate if it's registered (1) or not (0). 
I found that following invariant needs to be satisfied:

- For a MA, the flag should not be both 1 and 0 for different Children. MA with both flag 1 and 0 **should not** be included in the `MLR.req.`  If the flag can be both 1 and 0 on different Children, when sending `MLR.req`, for each MA with flag = 0, we will have to check if has flag 1 on other children. This lead to a time complexity of `O(TotalIpNum*TotalIpNum)`, which I think is something we should avoid.

This lead to the further consideration that when a Child adds a MA, we need to check all MAs on all Children to check if this MA is indeed newly listened. If this MA was registered (with flag=1) by the any Child, the MA's flag should be set to 1. If the MA was not listened or was listened but not registered (flag = 0), the MA's flag should be set to 0. 

Thus for time complexity, we can have folowings:
- Adding a MA to Child has a time complexity of O(TotalIpNum) to search for same MA with flag=1
- Sending MLR.req has a time complexity of O(TotalIpNum * IP6AddressesTlvSize). 

This however has not taken into consideration the bit mask optimization tricks @jwhui  have suggested. 
- To efficiently searching for MAs with flag=0, we can have a `mToRegisterMask` bit-vector for each Child. 
- To efficiently searching for MAs with flag=1, we can have a `mRegisteredMask` bit-vector for each Child. 

These bit masks will allow us to skip Child that has no flag=0 or flag=1 MAs.

Reply:
I think the important point is we want to quickly identify the set of unique multicast addresses. My proposal was to have a bit for each address that indicates whether the given multicast address is "authoritative" or "duplicate". When sending MR.req message, we only consider and maintain state for "authoritative" IPv6 addresses. This should avoid the O(N^2) time complexity and result in O(N) time complexity. Of course, it means that adding/removing a multicast address also has O(N) time complexity.

The bit-vector simply makes it a bit more efficient by avoiding the walk across all child address entries, but it's still O(N) time complexity.

Thoughts?

Reply:
![image](https://user-images.githubusercontent.com/52448427/87767276-ff07ea80-c84c-11ea-9f24-fc62e5cf7263.png)

Now that the proxying is more or less implemented, we can see the size change to compare the current `table-less` approach (in green) with the earlier `table` approach (in red).

- The `table-less` implemented is MTD-friendly to have less code size and RAM footprint. 
- For FTD, the `table-less` approaches does have code size advantage nor disadvantage but it can proxy more Child addresses (30 v.s. 15) with same RAM footprint.